File: backend/app/schemas/intelligence.py
"""
Pydantic schemas for Intelligence Hub API responses
"""
from pydantic import BaseModel, Field
from typing import List, Optional, Dict, Any
from datetime import datetime


# =========================================================================
# ENTITY SCHEMAS
# =========================================================================

class EntityResponse(BaseModel):
    """Basic entity response"""
    id: str
    canonical_name: str
    entity_type: str
    mention_count: int
    synthesis_count: int = 0
    as_cause_count: int = 0
    as_effect_count: int = 0
    topics: List[str] = []

    class Config:
        from_attributes = True


class EntityDetailResponse(EntityResponse):
    """Detailed entity response with additional info"""
    description: str = ""
    aliases: List[str] = []
    first_seen: Optional[datetime] = None
    last_seen: Optional[datetime] = None
    related_entities: List[str] = []
    synthesis_ids: List[str] = []


class EntitySearchResult(BaseModel):
    """Entity search result with similarity score"""
    entity: EntityResponse
    similarity_score: float


class EntityCausalProfile(BaseModel):
    """Entity's causal profile across syntheses"""
    entity_name: str
    entity_type: str = "UNKNOWN"
    as_cause: List[Dict[str, Any]] = []
    as_effect: List[Dict[str, Any]] = []
    total_causal_mentions: int = 0
    cause_ratio: float = 0.5


# =========================================================================
# TOPIC SCHEMAS
# =========================================================================

class TopicResponse(BaseModel):
    """Basic topic response"""
    id: str
    name: str
    description: str = ""
    category: str
    synthesis_count: int = 0
    entity_count: int = 0
    narrative_arc: str = "emerging"
    days_tracked: int = 0
    is_active: bool = True
    last_updated: Optional[datetime] = None
    preview_entities: List[str] = []

    class Config:
        from_attributes = True


class TopicDetailResponse(TopicResponse):
    """Detailed topic response"""
    keywords: List[str] = []
    synthesis_ids: List[str] = []
    entity_ids: List[str] = []
    first_seen: Optional[datetime] = None
    hot_score: float = 0.0


class TimelineLayer(BaseModel):
    """A layer in the timeline visualization"""
    period: str
    start_time: float
    end_time: float
    node_ids: List[str] = []
    edge_ids: List[str] = []


class CausalNode(BaseModel):
    """A node in the causal graph"""
    id: str
    label: str
    node_type: str = "event"
    fact_density: float = 0.5
    mention_count: int = 1
    source_syntheses: List[str] = []


class CausalEdge(BaseModel):
    """An edge in the causal graph"""
    id: str
    cause_text: str
    effect_text: str
    relation_type: str = "causes"
    confidence: float = 0.5
    mention_count: int = 1
    source_syntheses: List[str] = []


class TopicCausalGraphResponse(BaseModel):
    """Aggregated causal graph for a topic"""
    topic_id: str
    topic_name: str
    nodes: List[CausalNode] = []
    edges: List[CausalEdge] = []
    timeline_layers: List[TimelineLayer] = []
    central_entities: List[str] = []
    narrative_arc: str = "emerging"
    total_syntheses: int = 0


class TopicTimelineResponse(BaseModel):
    """Timeline of syntheses in a topic"""
    topic_id: str
    topic_name: str
    syntheses: List[Dict[str, Any]] = []
    timeline_layers: List[TimelineLayer] = []


# =========================================================================
# GLOBAL GRAPH SCHEMAS
# =========================================================================

class GlobalGraphNode(BaseModel):
    """A node in the global graph (topic or entity)"""
    id: str
    label: str
    node_type: str  # "topic" or "entity"
    weight: int = 1
    category: str = ""


class GlobalGraphEdge(BaseModel):
    """An edge in the global graph"""
    source: str
    target: str
    weight: float = 1.0
    edge_type: str = "related"


class GlobalGraphResponse(BaseModel):
    """Global overview graph of all topics/entities"""
    nodes: List[GlobalGraphNode] = []
    edges: List[GlobalGraphEdge] = []
    stats: Dict[str, Any] = {}


# =========================================================================
# LIST RESPONSES
# =========================================================================

class TopicsListResponse(BaseModel):
    """List of topics response"""
    topics: List[TopicResponse]
    total: int
    page: int = 1
    limit: int = 20


class EntitiesListResponse(BaseModel):
    """List of entities response"""
    entities: List[EntityResponse]
    total: int
    page: int = 1
    limit: int = 50


# =========================================================================
# STATS SCHEMAS
# =========================================================================

class IntelligenceStats(BaseModel):
    """Overall Intelligence Hub statistics"""
    total_topics: int = 0
    active_topics: int = 0
    total_entities: int = 0
    entities_by_type: Dict[str, int] = {}
    topics_by_category: Dict[str, int] = {}
    topics_by_arc: Dict[str, int] = {}

File: app/types/intelligence.ts
/**
 * TypeScript types for Intelligence Hub
 */

// =========================================================================
// ENTITY TYPES
// =========================================================================

export interface EntityResponse {
  id: string;
  canonical_name: string;
  entity_type: string;
  mention_count: number;
  synthesis_count: number;
  as_cause_count: number;
  as_effect_count: number;
  topics: string[];
}

export interface EntityDetailResponse extends EntityResponse {
  description: string;
  aliases: string[];
  first_seen: string | null;
  last_seen: string | null;
  related_entities: string[];
  synthesis_ids: string[];
}

export interface EntitySearchResult {
  entity: EntityResponse;
  similarity_score: number;
}

export interface EntityCausalProfile {
  entity_name: string;
  entity_type: string;
  as_cause: Array<{
    effect: string;
    confidence: number;
    synthesis_id: string;
    synthesis_title: string;
  }>;
  as_effect: Array<{
    cause: string;
    confidence: number;
    synthesis_id: string;
    synthesis_title: string;
  }>;
  total_causal_mentions: number;
  cause_ratio: number;
}

// =========================================================================
// TOPIC TYPES
// =========================================================================

export interface TopicResponse {
  id: string;
  name: string;
  description: string;
  category: string;
  synthesis_count: number;
  entity_count: number;
  narrative_arc: 'emerging' | 'developing' | 'peak' | 'declining' | 'resolved';
  days_tracked: number;
  is_active: boolean;
  last_updated: string | null;
  preview_entities: string[];
}

export interface TopicDetailResponse extends TopicResponse {
  keywords: string[];
  synthesis_ids: string[];
  entity_ids: string[];
  first_seen: string | null;
  hot_score: number;
}

// =========================================================================
// CAUSAL GRAPH TYPES
// =========================================================================

export interface TimelineLayer {
  period: string;
  start_time: number;
  end_time: number;
  node_ids: string[];
  edge_ids: string[];
}

export interface CausalNode {
  id: string;
  label: string;
  node_type: string;
  fact_density: number;
  mention_count: number;
  source_syntheses: string[];
}

export interface CausalEdge {
  id: string;
  cause_text: string;
  effect_text: string;
  relation_type: string;
  confidence: number;
  mention_count: number;
  source_syntheses: string[];
}

export interface TopicCausalGraphResponse {
  topic_id: string;
  topic_name: string;
  nodes: CausalNode[];
  edges: CausalEdge[];
  timeline_layers: TimelineLayer[];
  central_entities: string[];
  narrative_arc: string;
  total_syntheses: number;
}

export interface TopicTimelineResponse {
  topic_id: string;
  topic_name: string;
  syntheses: Array<{
    id: string;
    title: string;
    created_at: string;
    summary: string;
  }>;
  timeline_layers: TimelineLayer[];
}

// =========================================================================
// GLOBAL GRAPH TYPES
// =========================================================================

export interface GlobalGraphNode {
  id: string;
  label: string;
  node_type: 'topic' | 'entity';
  weight: number;
  category: string;
}

export interface GlobalGraphEdge {
  source: string;
  target: string;
  weight: number;
  edge_type: string;
}

export interface GlobalGraphResponse {
  nodes: GlobalGraphNode[];
  edges: GlobalGraphEdge[];
  stats: {
    total_topics: number;
    total_entities: number;
    total_syntheses: number;
  };
}

// =========================================================================
// LIST RESPONSES
// =========================================================================

export interface TopicsListResponse {
  topics: TopicResponse[];
  total: number;
  page: number;
  limit: number;
}

export interface EntitiesListResponse {
  entities: EntityResponse[];
  total: number;
  page: number;
  limit: number;
}

// =========================================================================
// STATS TYPES
// =========================================================================

export interface IntelligenceStats {
  total_topics: number;
  active_topics: number;
  total_entities: number;
  entities_by_type: Record<string, number>;
  topics_by_category: Record<string, number>;
  topics_by_arc: Record<string, number>;
}
